File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int res = 0;
        int low = INT_MAX;

        for (int price : prices) {
            if (price < low) {
                low = price;
            }

            if (price - low > res) {
                res = price - low;
            }
        }

        return res;
        
    }
};

File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.java
class Solution {
    public int maxProfit(int[] prices) {
        int[][] dp = new int[prices.length][2];

        dp[0][0] = 0;
        dp[0][1] = -prices[0];

        for (int i = 1; i < prices.length; i++) {
            dp[i][0] = Math.max(dp[i - 1][0], dp[i - 1][1] + prices[i]);
            dp[i][1] = Math.max(dp[i - 1][1], -prices[i]);
        }

        return Math.max(dp[prices.length - 1][0], dp[prices.length - 1][1]);


        // int min = Integer.MAX_VALUE;
        // int max = 0;

        // for (int i = 0; i < prices.length; i++) {
        //     if (prices[i] < min) {
        //         min = prices[i];
        //     }

        //     if (prices[i] - min > max) {
        //         max = prices[i] - min;
        //     }
        // }

        // return max;   
    }
}